docs/madr/decisions/049-meshexternalservice-api.md

```diff
@@ -0,0 +1,518 @@
+
+```yaml
+kind: HostnameGenerator
+metadata:
+  name: k8s-zone-hostnames
+spec:
+  targetRef:
+    kind: MeshExternalService
+    tags:
+      my.label.io/service: my-service 
+  template: {{ name }}.svc.meshext.local
+```
+
+For example, given this `MeshExternalService`
+
+```yaml
+kind: MeshExternalService
+metadata:
+  name: mydomain
+  namespace: kuma-system
+  labels:
+    my.label.io/service: my-service 
+    kuma.io/mesh: default
+    kuma.io/zone: east-1
+    kuma.io/origin: zone
+spec:
+  destinations:
+  - address: 192.168.0.1
+    port: 9090 
+...
+```
+
+We would generate such hostname `mydomain.svc.meshext.local`
+
+Possible template functions/keys are:
+
+`{{ name }}` - name of the `MeshExternalService`
+`{{ label "x" }}` - value of label x.
+
+If the template cannot be resolved (label is missing), the hostname won't be generated with a NotAvailable status and an appropriate reason.
+
+#### Positive Consequences
+
+* Simpler model, but not as clear
+* One domain per external service 
+
+#### Negative Consequences
+
+* Using HostnameGenerator might cause unexpected domains to be created, when match is too wide
+
+#### Creating MeshExternalService and using HostnameGenerator for InternalVIP
+
+This option is similar to the chosen one, but instead of explicitly providing a domain in the policy, users would need to create a `HostnameGenerator` to generate a domain.
+
+The API model does not include a `value` field for the `InternalVIP` type. The domain name is generated by `HostnameGenerator` and later provided under a status section.
+
+```yaml
+kind: MeshExternalService
+metadata:
+  name: example
+  namespace: kuma-system
+  labels:
+    kuma.io/mesh: default
+    kuma.io/zone: east-1
+spec:
+  match:
+  - type: InternalVIP # Kuma will generate a domain based on HostnameGenerator
+    port: 80
+    protocol: http
+...
+status:
+  addresses:
+  - hostname: httpbin.svc.meshext.local
+    status: Available # | NotAvailable
+    origin: 
+      kind: HostGenerator
+      name: "k8s-zone-generator"
+    reason: "not available because of the clash with ..."
+  vips:
+  - ip: <kuma VIP>
+    type: Kuma | # External
+```
+
+#### Positive Consequences
+
+* Separation between hostname generation and possible better security (but not a huge problem when only MeshOperator is allowed to create the policy)
+
+#### Negative Consequences
+
+* Less explicit
+* Match section looks a bit awkward with a hostname generator
+
+## Common concerns
+
+### Migration
+
+It is vital that we provide easy and safe migration path.
+MeshExternalService is quite different from other policies so having shadow mode will result in big Envoy config diffs that would be hard to predict if they would work correctly.
+
+#### Introduce switch on the DP (env variable / label) to disable specific ExternalServices
+
+We can have a switch to disable a list of ExternalServices on the DP level.
```

Comment:
How about using `kuma.io/ignore` https://kuma.io/docs/dev/reference/kubernetes-annotations/#kumaioignore ?

How do we define overlap?

Reply:
We can extend it for ExternalService. The only issue is that we would need something else if we would like to disable ExternalService only for specific DP. One option is defining reachable services, but if a user uses all services, it is difficult to list all services except one.

Reply:
We can start with `kuma.io/ignore`. What do you mean by overlap?